gazelle_haskell_modules/rule_generation.go

```diff
@@ -265,21 +274,62 @@ func concatRuleInfos(xs [][]*RuleInfo) []*RuleInfo {
 	return ys
 }
 
-// Collects the source files referenced in the given expression
-func srcsFromRule(pkgRoot string, expr build.Expr) ([]string, error) {
-	srcs, err := getSources(expr)
+// Find and append together
+// 0. The files listed in the srcs attribute of a rule
+// 1. All the files (recursively) in directories specified in a "gazelle_haskell_modules:srcs:"
+func getSrcs(pkgRoot string, r *rule.Rule) ([]string, error) {
+	sourcesFromRule, err := srcsFromRule(pkgRoot, r)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("getSrcs: %w", err)
 	}
 
-	xs := make([]string, len(srcs))
-	i := 0
-	for f, _ := range srcs {
-		xs[i] = path.Join(pkgRoot, f)
-		i++
+	autodiscoveredSrcs, err := autodiscoverSrcs(pkgRoot, r)
+	if err != nil {
+		return nil, fmt.Errorf("getSrcs: %w", err)
 	}
 
-	return xs, nil
+	// reuse the set we got form srcsFromRule to deduplicate autodiscovered and explicit files
+	for _, v := range autodiscoveredSrcs {
+		sourcesFromRule[v] = true
+	}
+
+	var srcs []string
+	for file := range sourcesFromRule {
+		srcs = append(srcs, file)
+	}
+	return srcs, nil
+}
+
+func srcsFromRule(pkgRoot string, r *rule.Rule) (map[string]bool, error) {
+	if expr := r.Attr("srcs"); expr != nil {
+		files, err := getStringList(expr)
+		if err != nil {
+			return nil, fmt.Errorf("srcsFromRule: %w", err)
+		}
+
+		sourceMap := make(map[string]bool, len(files))
+		for _, file := range files {
+			sourceMap[path.Join(pkgRoot, file)] = true
+		}
+
+		return sourceMap, nil
+	}
+	return map[string]bool{}, nil
+}
+
+func autodiscoverSrcs(pkgRoot string, r *rule.Rule) ([]string, error) {
+
+	srcDirs, err := getSrcDirsFromRuleDirective(r)
```

Comment:
```suggestion
	srcDirs, err := getSrcDirsFromComments(r.Comments())
```
Looks like this change would get rid of the wrapping function `getSrcDirsFromRuleDirective`.

I don't feel strong about it, so feel free to ignore it.

Reply:
Ah, of course, thanks :sweat_smile: